scripts/publish.js

```diff
@@ -0,0 +1,265 @@
+
+import {
+	getPackageDirs,
+	calculatePackageDependencies,
+	createBuildOrder,
+} from "./shared.js";
+import { execSync } from "node:child_process";
+import { existsSync, readFileSync } from "node:fs";
+import { join } from "node:path";
+
+//-----------------------------------------------------------------------------
+// Read CLI Args
+//-----------------------------------------------------------------------------
+
+const dryRun = process.argv.includes("--dry-run");
+
+// for dry runs only output to console and don't execute anything
+const exec = dryRun ? text => console.log(text) : execSync;
+
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+/**
+ * Converts a GitHub Actions step output name to its corresponding environment variable name
+ * @param {string} stepId The ID of the step
+ * @param {string} outputName The name of the output
+ * @returns {string} The environment variable name
+ */
+function convertOutputToEnvVar(stepId, outputName) {
+	// Convert step ID to uppercase and replace hyphens with underscores
+	const normalizedStepId = stepId.toUpperCase().replace(/-/gu, "_");
+
+	// Convert output name to uppercase and replace slashes with double underscores
+	const normalizedOutputName = outputName
+		.toUpperCase()
+		.replace(/\//gu, "__")
+		.replace(/-/gu, "_");
+
+	return `STEPS_${normalizedStepId}_OUTPUTS_${normalizedOutputName}`;
+}
+
+/**
+ * Gets the output of a GitHub Actions step from the environment variables.
+ * @param {string} packageDir The directory of the package.
+ * @param {string} name The name of the output.
+ * @return {string} The output value.
+ */
+function getReleaseOutput(packageDir, name) {
+	return process.env[
+		convertOutputToEnvVar("release", `${packageDir}--${name}`)
+	];
+}
+
+/**
+ * Gets the list of packages to publish.
+ * @param {Array<string>} packageDirs The list of package directories.
+ * @return {Array<string>} The list of packages to publish.
+ */
+function getPackagesToPublish(packageDirs) {
+	return packageDirs.filter(
+		packageDir =>
+			getReleaseOutput(packageDir, "release_created") === "true",
+	);
+}
+
+/**
+ * Maps the dependencies into a structure where they keys are package
+ * paths and the values are an array of package paths.
+ * @param {Map<string, { name:string, dir: string, dependencies: Set<string> }>} dependencies The dependencies to map.
+ * @return {Map<string, Set<string>>} The mapped dependencies.
+ */
+function mapDependenciesToPaths(dependencies) {
+	const mappedDependencies = new Map();
+
+	for (const [, { dir: packageDir, dependencies: deps }] of dependencies) {
+		const pathDeps = [...deps]
+			.filter(dep => dependencies.has(dep))
+			.map(dep => {
+				const depDir = dependencies.get(dep);
+				if (depDir) {
+					return depDir.dir;
+				}
+				return dep;
+			});
```

Comment:
```suggestion
			.map(dep => dependencies.get(dep).dir);
```

Can this be simplified this way? Because the array is already filtered by `dependencies.has(dep)`.